Merges the changes from the `release/3.0` branch to `release/4.0` branch

nikcio opened the pull request:
This merges the changes in the `release/3.0` branch to the `release/4.0` branch. This primarily contains #320.

@nzdev Is there any way we can use faceting and the feature in #320 at the same time? I couldn't see an obvious way to accomplice this.

And @nzdev What would be a good XML comment for `ShardIndex` in the `src/Examine.Lucene/Search/LuceneSearchResult.cs` file?

nzdev commented:
https://github.com/Shazwazza/Examine/pull/321 is facet deep paging

nzdev commented:
Here's what I'm thinking.
Have the next release of Examine be 4.0, but avoid breaking API changes for 3.x. This means Umbraco 10 and 12 can choose to relax the allowed Examine version to be v3 or v4.

Steps:
1. Merge PR https://github.com/Shazwazza/Examine/pull/346 which tracks the shipped API for V3.
2. Merge PR https://github.com/Shazwazza/Examine/pull/347 which merges V3 into V4 and fixes any API compatibility issues.
3. Rebase https://github.com/Shazwazza/Examine/pull/345 to fix any nullability / xml docs issues.
4. Add the new api's as unshipped to the txt files and release a beta
5. Allow time for feedback, resolve feedback, add new api to the shipped.txt files.
6. Release 4.0

Shazwazza commented:
Looks like conflicts need to be resolved

nikcio commented on src/Examine.Lucene/Providers/LuceneIndex.cs:
```diff
@@ -49,6 +59,18 @@ public class LuceneIndex : BaseIndexProvider, IDisposable, IIndexStats, Referenc
             //initialize the field types
             _fieldValueTypeCollection = new Lazy<FieldValueTypeCollection>(() => CreateFieldValueTypes(_options.IndexValueTypesFactory));
 
+            if (_options.UseTaxonomyIndex)
+            {
+                _taxonomySearcher = new Lazy<LuceneTaxonomySearcher>(CreateTaxonomySearcher);
+
+                _searcher = new Lazy<BaseLuceneSearcher>(() => _taxonomySearcher.Value);
+            }
+            else
+            {
+                _taxonomySearcher = new Lazy<LuceneTaxonomySearcher>(() => throw new NotSupportedException("TaxonomySearcher not supported when not using taxonomy index."));
+                _searcher = new Lazy<BaseLuceneSearcher>(CreateSearcher);
+            }
+
```
@nzdev If I remember correctly you added this protected constructor. It was complaining that the `_searcher` wasn't set. So I've added the same code as from the private constructor. Do you believe this is the right course of action?

nzdev replied:
Looks good to me.

nikcio commented on src/Examine.Lucene/Search/LuceneSearchExecutor.cs:
```diff
@@ -243,10 +242,6 @@ private static SearchAfterOptions GetSearchAfterOptions(TopDocs topDocs)
 
             var facetFields = _facetFields.OrderBy(field => field.FacetField);
 
-            SortedSetDocValuesReaderState sortedSetReaderState = null;
-            Facets fastTaxonomyFacetCounts = null;
-
-
```
Wasn't used

nikcio commented on src/Examine.Lucene/Providers/LuceneSearcher.cs:
```diff
@@ -51,8 +51,23 @@ public LuceneSearcher(string name, SearcherManager searcherManager, Analyzer ana
         public override ISearchContext GetSearchContext()
             => new SearchContext(_searcherManager, _fieldValueTypeCollection);
 
+        //protected override void Dispose(bool disposing)
+        //{
+        //if (!_disposedValue)
+        //{
+        //    if (disposing)
+        //    {
+        //        _searcherManager.Dispose();
+        //    }
+
+        //    _disposedValue = true;
+        //}
+        //base.Dispose(disposing);
+        //}
```
This will replace the `new virtual` when possible to align with the dispose pattern.

nzdev replied:
Looks like this would break any classes overriding public virtual void Dispose(). We should revert to keep compatibility and mark as obsolete in V5.

nikcio replied:
This code is commented out I can't see it being problematic before uncommented?

nikcio commented on src/Examine.Lucene/Providers/BaseLuceneSearcher.cs:
```diff
@@ -95,15 +98,26 @@ public override ISearchResults Search(string searchText, QueryOptions? options =
+
+        /// <summary>
+        /// Disposes of the searcher
+        /// </summary>
+        /// <param name="disposing"></param>
+        protected virtual void Dispose(bool disposing)
         {
-            return new FacetsConfig();
+            if (!_disposedValue)
+            {
+                _disposedValue = true;
+            }
         }
 
-        /// <inheritdoc/>
-        public virtual void Dispose()
+        /// <inheritdoc />
+        public void Dispose()
         {
-
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
```
Align with the dispose pattern

nzdev replied:
Looks like this would break any classes overriding public virtual void Dispose(). We should revert to keep compatibility and mark as obsolete in V5.

nikcio replied:
@nzdev I've readded the `virtual` modifier and added an `Obsolete` attribute describing that you should override the `Dispose(bool)` instead.

nikcio commented:
@Shazwazza This is now updated. I've fixed all warnings that isn't part of the `RSxxx` warnings and had a simple fix. Above I've added some comments to the most important points of interest.

nzdev commented:
RS0036 should still be active. The unshipped API .txts files should get deleted and regenerated now that nullable / API should be ready. Will review rest later

nikcio commented:
@nzdev super then we can remove the nowarn

nikcio commented:
>RS0036 should still be active. The unshipped API .txts files should get deleted and regenerated now that nullable / API should be ready. Will review rest later
@nzdev I tried generating the files but couldn't seem to get all the warnings to go away may be I'm doing something incorrectly.

But I think the `Directory.Build.props` should have this configuration at the bottom:
```xml
  <!-- Disable the nullable warnings when compiling for .NET Standard 2.0 -->
  <PropertyGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
    <NoWarn>$(NoWarn);nullable;RS0037</NoWarn>
  </PropertyGroup>

  <ItemGroup Condition="'$(MSBuildProjectName)' != 'Examine.Web.Demo' AND '$(MSBuildProjectName)' != 'Examine.Test'">
    <AdditionalFiles Include="PublicAPI/$(TargetFramework)/PublicAPI.Shipped.txt" />
    <AdditionalFiles Include="PublicAPI/$(TargetFramework)/PublicAPI.Unshipped.txt" />
  </ItemGroup>
```

I think we need to ignore `RS0037` on `netstandard2.0` because we don't fully support nullable on this target and for the same reason I think we need to have a PublicAPI file for each of the target frameworks. What do you think and could you help me with generating the PublicAPI files correctly?

nzdev commented:
https://github.com/Shazwazza/Examine/pull/349 addresses the public API files and fixes / suppresses most of the warnings.

nikcio commented:
#349 should include everything from this one so i'll close this one
